src/tools/attack/attack.cpp

```diff
@@ -106,31 +101,55 @@ UdpUnreachCallback(
 {
 }
 
-void RunAttackRandom(CXPLAT_SOCKET* Binding, uint16_t Length, bool ValidQuic)
+static
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Function_class_(CXPLAT_ROUTE_RESOLUTION_CALLBACK)
+void
+ResolveRouteComplete(
+    _Inout_ void* Context,
+    _When_(Succeeded == FALSE, _Reserved_)
+    _When_(Succeeded == TRUE, _In_reads_bytes_(6))
+        const uint8_t* PhysicalAddress,
+    _In_ uint8_t PathId,
+    _In_ BOOLEAN Succeeded
+    )
 {
-    CXPLAT_ROUTE Route = {0};
-    CxPlatSocketGetLocalAddress(Binding, &Route.LocalAddress);
-    Route.RemoteAddress = ServerAddress;
+    UNREFERENCED_PARAMETER(PathId);
+    if (Succeeded) {
+        CxPlatResolveRouteComplete(nullptr, (CXPLAT_ROUTE*)Context, PhysicalAddress, 0);
+    }
+}
 
+//========================================Attacks========================================
+void RunAttackFlooding(CXPLAT_SOCKET* Binding, uint16_t Length, bool ValidQuic, bool TCP = false)
+{
     uint64_t ConnectionId = 0;
     CxPlatRandom(sizeof(ConnectionId), &ConnectionId);
 
-    while (CxPlatTimeDiff64(TimeStart, CxPlatTimeMs64()) < TimeoutMs) {
+    CXPLAT_ROUTE Route = {0};
+    CxPlatSocketGetLocalAddress(Binding, &Route.LocalAddress);
+    CxPlatSocketGetRemoteAddress(Binding, &Route.RemoteAddress);
 
-        CXPLAT_SEND_CONFIG SendConfig = { &Route, Length, CXPLAT_ECN_NON_ECT, 0 };
-        CXPLAT_SEND_DATA* SendData = CxPlatSendDataAlloc(Binding, &SendConfig);
-        if (SendData == nullptr) {
-            printf("CxPlatSendDataAlloc failed\n");
+    if (CxPlatIsRawInitialized(Datapath))
+    {
+        if (QUIC_FAILED(CxPlatResolveRoute(Binding, &Route, 0, &Route, ResolveRouteComplete))) {
```

Comment:
This is asynchronous. You need to wait on it.

Reply:
Done.